cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -481,88 +510,84 @@ private void addArchivesToList(List<ArtifactDetail> archives, Row row) throws IO
     }
   }
 
-  // given a row representing plugin metadata, add all plugins in the row to the given map
-  private void addPluginsToMap(Map<ArtifactInfo, List<PluginClass>> map, Row row) throws IOException {
-    // plugin key contains namespace, plugin type, and plugin name
-    PluginKey pluginKey = PluginKey.parse(row.getRow());
-
+  private void addPluginsToMap(Id.Artifact parentArtifactId, Map<ArtifactInfo, List<PluginClass>> map,
+                               Row row) throws IOException {
     // column is the artifact name and version, value is the serialized PluginClass
     for (Map.Entry<byte[], byte[]> column : row.getColumns().entrySet()) {
       ArtifactColumn artifactColumn = ArtifactColumn.parse(column.getKey());
       PluginData pluginData = gson.fromJson(Bytes.toString(column.getValue()), PluginData.class);
 
-      Id.Artifact artifactId =
-        Id.Artifact.from(pluginKey.namespace, artifactColumn.name, artifactColumn.version);
-      ArtifactInfo artifactInfo = new ArtifactInfo(artifactId, pluginData.artifactLocation);
+      // filter out plugins that don't extend this version of the parent artifact
+      if (matches(pluginData, parentArtifactId.getVersion())) {
```

Comment:
we don't need to check namespace and name because the scan will filter out anything that doesn't match.  But there isn't a way to do something similar with a version range, at least none that I could think of.